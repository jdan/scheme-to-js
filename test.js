const { spawn } = require("child_process")
const prettier = require("prettier")

function compile(code, cb) {
  const scheme = spawn("scheme", ["-q"])

  let output = ""
  scheme.stdout.on("data", data => {
    output += data
  })
  scheme.on("close", () => {
    cb(prettier.format(output, { parser: "babylon" }))
  })

  scheme.stdin.end(`
    (load "compiler.scm")
    (display (scheme->js ${code}))
  `)
}

describe("+", () => {
  it("should work", done => {
    compile(`(+ 3 4)`, output => {
      expect(output).toMatchSnapshot()
      done()
    })
  })
})

describe("big block of code from main", () => {
  it("should match snapshots", done => {
    compile(
      `
      (begin
        (define (string-join strs joiner)
            (define (helper strs acc)
                (if (null? strs)
                    acc
                    (helper (cdr strs)
                            (+ acc
                               (+ joiner (car strs))))))
            (if (null? strs)
                ""
                (helper (cdr strs) (car strs))))
        (string-join (Array "apples" "bananas" "cucumbers") ","))))
    `,
      output => {
        expect(output).toMatchSnapshot()
        expect(eval(output)).toMatchSnapshot()
        done()
      }
    )
  })
})
