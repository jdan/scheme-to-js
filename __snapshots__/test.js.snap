// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`+ should work 1`] = `
"3 + 4;
"
`;

exports[`big block of code from main should match snapshots 1`] = `
"(() => {
  function null$(ls) {
    return 0 === ls.length;
  }
  function car(ls) {
    return Array.prototype.shift.call(ls, Array.prototype.slice.call(ls));
  }
  function cdr(ls) {
    return Array.prototype.slice.call(ls, 1);
  }
  function string$append(a, b, c) {
    return a + (b + c);
  }
  function string$join(strs, joiner) {
    return (() => {
      function helper(strs, acc) {
        return null$(strs)
          ? acc
          : helper(cdr(strs), string$append(acc, joiner, car(strs)));
      }
      return null$(strs) ? \\"\\" : helper(cdr(strs), car(strs));
    })();
  }
  return string$join(Array(\\"apples\\", \\"bananas\\", \\"cucumbers\\"), \\",\\");
})();
"
`;

exports[`big block of code from main should match snapshots 2`] = `"apples,bananas,cucumbers"`;
